core/container.go

```diff
@@ -268,6 +269,43 @@ func (container *Container) Mounts(ctx context.Context) ([]string, error) {
 	return mounts, nil
 }
 
+func (container *Container) Directory(ctx context.Context, dir string) (*Directory, error) {
+	payload, err := container.ID.decode()
+	if err != nil {
+		return nil, err
+	}
+
+	// NB(vito): iterate in reverse order so we'll find deeper mounts first
+	for i := len(payload.Mounts) - 1; i >= 0; i-- {
```

Comment:
There are some nightmare-inducing corner cases involving symlinks here I think (double check my thinking though). I'm inclined to not care about them right now because I'm not sure users will, but they are worth being aware of:
1. Rootfs has a symlink `/link -> /target`
2. Use our api to put a mount at `/link/mnt`. This should result in the mount really being at `/target/mnt` during exec ops (I *think* buildkit handles that via runc/containerd). So your code will see that `/target/mnt/dir` is a valid, existing directory (and may output it to stdout).
3. Call this api with `directory(dir: "/target/mnt/dir")` and you will not find the dir you may have been expecting because the mount was created with the API at `/link`, not `/target`. You may get some entirely different dir.

Like I said, my hope is that these cases are so obscure we don't have to care about them right now, but I could be wrong (people have gotten to this level of detail when asking about buildkit+dockerfiles). The only possible solutions are too complicated to be worth it right now I think. Bringing it up anyways in case anyone thinks of any related issues that are less obscure and more immediately pressing.

Reply:
Oh yeah hm that sounds annoying.

It's really tempting to just say "don't do that." In general I would expect the reader of the GraphQL API query to be able to recognize what `directory(dir)` is trying to access, i.e. a previously configured mount. It's strange to me to say "mount this at X but then read it from Y" in one API call.

But we'll have to see. Users always have the best surprises.

Reply:
Yeah totally, I do think the risk is gonna be increased when users are getting `Container` returned from other extensions, so they don't see the full chain and original mount paths. But still likely obscure enough to not care right now. I made an issue so we don't forget about it, but not high priority: https://github.com/dagger/dagger/issues/3269